src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/DeviceTest.cs

```diff
@@ -0,0 +1,145 @@
+using NUnit.Framework;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Xml;
+using System.Xml.Linq;
+using System.Xml.XPath;
+using Xamarin.ProjectTools;
+using XABuildPaths = Xamarin.Android.Build.Paths;
+
+namespace Xamarin.Android.Build.Tests
+{
+	public class DeviceTest: BaseTest
+	{
+		protected static void RunAdbInput (string command, params object [] args)
+		{
+			RunAdbCommand ($"shell {command} {string.Join (" ", args)}");
+		}
+
+		protected static string ClearAdbLogcat ()
+		{
+			return RunAdbCommand ("logcat -c");
+		}
+
+		protected static string ClearDebugProperty ()
+		{
+			return RunAdbCommand ("shell setprop debug.mono.extra \"\"");
+		}
+
+		protected static void AdbStartActivity (string activity)
+		{
+			RunAdbCommand ($"shell am start -S -n \"{activity}\"");
+		}
+
+		protected void WaitFor (TimeSpan timeSpan, Func<bool> func)
+		{
+			var pause = new ManualResetEvent (false);
+			TimeSpan total = timeSpan;
+			TimeSpan interval = TimeSpan.FromMilliseconds (10);
+			while (total.TotalMilliseconds > 0) {
+				pause.WaitOne (interval);
+				total = total.Subtract (interval);
+				if (func ()) {
+					break;
+				}
+			}
+		}
+
+		protected static bool MonitorAdbLogcat (Func<string, bool> action, int timeout = 10)
+		{
+			string ext = Environment.OSVersion.Platform != PlatformID.Unix ? ".exe" : "";
+			string adb = Path.Combine (AndroidSdkPath, "platform-tools", "adb" + ext);
+			var info = new ProcessStartInfo (adb, "logcat") {
+				RedirectStandardOutput = true,
+				RedirectStandardError = true,
+				UseShellExecute = false,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+			};
+			using (var proc = Process.Start (info)) {
+				var sw = new Stopwatch ();
+				try {
+					TimeSpan time = TimeSpan.FromSeconds (timeout);
+					while (time.TotalMilliseconds > 0) {
+						sw.Start ();
+						if (action (proc.StandardOutput.ReadLine ()))
+							return true;
+						time = time.Subtract (TimeSpan.FromMilliseconds (sw.ElapsedMilliseconds));
+						sw.Reset ();
```

Comment:
It feels like we may be "unfairly counting"  the `sw.Reset()` and `time.Subtract()` and `TimeSpan.FromMilliseconds()` calls, in that these will count "toward" the timeout value without being "part of" the actual process output generation or handling.

Though given that `timeout` is 10s, this likely doesn't matter.

Reply:
I don't think it matters, it doesn't needed to be millisecond accurate.